main.c

```diff
@@ -332,6 +332,12 @@ FILE * check_file(char * dir, char *fname)
 	FILE * fp = NULL;
 	char path[256];
 
+	// Prevent USB device from requesting files in parent directories
+	if(strstr(fname, ".."))
+	{
```

Comment:
What's the reason for this?  Can you just add an error message to explain why this is disabled?

Reply:
Patchset aims to make it safer to allow unprivileged users in multi-user setups run rpiboot, without effectively handing them the keys to the entire system.


``rpiboot`` essentially is file server software handing out whatever file any attached USB device claiming to be a Raspberry Pi asks for.

If I enter ``rpiboot /usr/share/rpiboot/msd`` I have the reasonable expectation that the file server code only shares my ``/usr/share/rpiboot/msd`` folder with the connected device.
However currently nothing prevents a device to ask for a file name like ``../../../../etc/shadow``, which will get translated to ``/usr/share/rpiboot/msd/../../../../etc/shadow`` and serve ``/etc/shadow``

Simply rejecting any request with a .. in it is the simplest form of protection against directory traversal attacks like this.
Could also turn the path provided into an absolute path first and verify if that is within the path allowed however that requires platform specific code.

Reply:
Correction it does allow enumeration without privileges.
Added a patch to print a message.

Reply:
It would be good to see the path traversal patch being merged soon, testing against a stock rpiboot as expected it happily sends "../../../../etc/shadow".

> Found serial number 1
Second stage boot server
Received message GetFileSize: ../../../../etc/shadow
File size = 885 bytes
Received message ReadFile: ../../../../etc/shadow
File read: ../../../../etc/shadow

And with the "Disallow requests for files outside directory" patch it's denied.

> Second stage boot server
Received message GetFileSize: ../../../../etc/shadow
Cannot open file ../../../../etc/shadow
Received message ReadFile: ../../../../etc/shadow
No file ../../../../etc/shadow found